src/runtime/tracing.cpp

```diff
@@ -10,7 +10,105 @@ typedef int32_t (*trace_fn)(void *, const halide_trace_event_t *);
 
 namespace Halide { namespace Runtime { namespace Internal {
 
-WEAK int halide_trace_file = 0;
+// A spinlock that allows for shared and exclusive access. It's
+// equivalent to a reader-writer lock, but in my case the "readers"
+// will actually be writing simultaneously to the trace buffer, so
+// that's a bad name. We use the __sync primitives used elsewhere in
+// the runtime for atomic work. They are well supported by clang.
+class SharedExclusiveSpinLock {
+    uint32_t lock;
+    const static uint32_t exclusive_held_mask = 0x80000000;
+    const static uint32_t exclusive_waiting_mask = 0x40000000;
+    const static uint32_t shared_mask = 0x3fffffff;
+
+public:
+    void acquire_shared() {
+        while (1) {
+            uint32_t x = lock & shared_mask;
+            if (__sync_bool_compare_and_swap(&lock, x, x + 1)) {
+                return;
+            }
+        }
+    }
+
+    void release_shared() {
+        __sync_fetch_and_sub(&lock, 1);
+    }
+
+    void acquire_exclusive() {
+        while (1) {
+            __sync_fetch_and_or(&lock, exclusive_waiting_mask);
+            if (__sync_bool_compare_and_swap(&lock, exclusive_waiting_mask, exclusive_held_mask)) {
+                return;
+            }
+        }
+    }
+
+    void release_exclusive() {
+        __sync_fetch_and_xor(&lock, exclusive_held_mask);
+    }
+
+    SharedExclusiveSpinLock() : lock(0) {}
+};
+
+class TraceBuffer {
+    SharedExclusiveSpinLock lock;
+    uint32_t cursor;
+    uint8_t buf[1024*1024];
+
+public:
+
+    // Attempt to atomically acquire space in the buffer to write a
+    // packet. Returns NULL if the buffer was full.
+    halide_trace_packet_t *try_acquire_packet(uint32_t size) {
```

Comment:
Really not a fan of functions where the lock state is not either constant precondition or post condition.

Reply:
I figured acquire, release, and try_acquire are the only flavors of function that may change lock state. This function is a try_acquire. But it's just a helper. I made it private so that the interface only exposes acquire, release, and flush.